04.ls/ls.rb

```diff
@@ -20,6 +53,77 @@ def prepare_files(options)
   options['r'] ? sorted_files.reverse : sorted_files
 end
 
+def prepare_file_details(files)
+  file_details = []
+  blocks = 0
+
+  files.each do |filename|
+    filestat = File.stat(filename)
+    blocks += filestat.blocks
+    file_details << build_file_detail(filename, filestat)
+  end
+  [blocks, file_details]
+end
+
+def build_file_detail(filename, filestat)
+  file_permissions = build_file_permissions(filestat)
+  {
+    permission: file_permissions,
+    link: filestat.nlink,
+    owner_name: Etc.getpwuid(filestat.uid).name,
+    group_name: Etc.getgrgid(filestat.gid).name,
+    size: filestat.size,
+    timestamp: filestat.mtime.strftime('%b %e %H:%M'),
+    name: filename
+  }
+end
+
+def build_file_permissions(filestat)
+  octal_mode = filestat.mode.to_s(8)
+
+  file_permissions = [FILE_TYPE[filestat.ftype]]
+  special_permission_bits = octal_mode.slice(-4, 1)
+
+  octal_mode.chars[-3..].each_with_index do |bit, idx|
+    permission = PERMISSIONS[bit]
+    special_flag = SPECIAL_FLAGS[special_permission_bits]
+    file_permissions << if special_flag && special_flag[:idx] == idx
+                          update_permission(special_flag, permission)
+                        else
+                          permission.values_at(:r, :w, :x).join
+                        end
+  end
+
+  file_permissions.join
+end
+
+def update_permission(special_flag, permission)
+  updated_execute =
+    permission[:x] == 'x' ? special_flag[:with_x] : special_flag[:without_x]
+
+  "#{permission[:r]}#{permission[:w]}#{updated_execute}"
+end
+
+def output_long_format(blocks, file_details)
+  puts "total #{blocks / 2}"
+  link_width, owner_width, group_width, size_width =
+    %i[link owner_name group_name size].map do |key|
+      file_details.map { |file_detail| file_detail[key].to_s.size }.max
+    end
+
+  file_details.each do |file_detail|
+    puts [
+      file_detail[:permission],
+      file_detail[:link].to_s.rjust(link_width),
+      file_detail[:owner_name].to_s.ljust(owner_width),
+      file_detail[:group_name].to_s.ljust(group_width),
+      file_detail[:size].to_s.rjust(size_width),
+      file_detail[:timestamp],
+      file_detail[:name]
+    ].join(' ')
```

Comment:
```suggestion
    hoge = [
      file_detail[:permission],
      file_detail[:link].to_s.rjust(link_width),
      file_detail[:owner_name].to_s.ljust(owner_width),
      file_detail[:group_name].to_s.ljust(group_width),
      file_detail[:size].to_s.rjust(size_width),
      file_detail[:timestamp],
      file_detail[:name]
    ]
    puts hoge.join(' ')
```

ちょっと一気にいろいろやり過ぎてるので、一回変数に入れましょう

Reply:
一度変数に入れる形に修正しました。
こちらの`hoge`の部分の変数名も迷ったのですが、ファイル情報を呼び出していることまで分かるので、シンプルに`row`にしてみました。